### ISSUE TITLE
Add support to the Alienware 13 R3 computer

### ISSUE THREAD

**Issue by BuckarewBanzai** [rsm-gh/akbl#17]
```
Device: ID 187c:0529 Alienware Corporation 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               0.02
  bDeviceClass            0 (Defined at Interface level)
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x187c Alienware Corporation
  idProduct          0x0529 
  bcdDevice            0.00
  iManufacturer           1 Alienware
  iProduct                2 AW13
  iSerial                 3 03.00
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength           41
    bNumInterfaces          1
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xe0
      Self Powered
      Remote Wakeup
    MaxPower                0mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 No Subclass
      bInterfaceProtocol      0 None
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.01
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      56
          Report Descriptor: (length is 56)
            Item(Global): Usage Page, data= [ 0x00 0xff ] 65280
                            (null)
            Item(Local ): Usage, data= [ 0x01 ] 1
                            (null)
            Item(Main  ): Collection, data= [ 0x01 ] 1
                            Application
            Item(Global): Usage Page, data= [ 0x0c ] 12
                            Consumer
            Item(Local ): Usage, data= [ 0x06 ] 6
                            (null)
            Item(Main  ): Collection, data= [ 0x02 ] 2
                            Logical
            Item(Local ): Usage Minimum, data= [ 0x00 ] 0
                            Unassigned
            Item(Local ): Usage Maximum, data= [ 0x66 ] 102
                            Still
            Item(Global): Logical Minimum, data= [ 0x00 ] 0
            Item(Global): Logical Maximum, data= [ 0x01 ] 1
            Item(Global): Report ID, data= [ 0x01 ] 1
            Item(Global): Report Size, data= [ 0x08 ] 8
            Item(Global): Report Count, data= [ 0x0b ] 11
            Item(Main  ): Input, data= [ 0x02 ] 2
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Bitfield
            Item(Main  ): End Collection, data=none
            Item(Global): Usage Page, data= [ 0x14 ] 20
                            Alphanumeric Display
            Item(Local ): Usage, data= [ 0x2b ] 43
                            Character Report
            Item(Main  ): Collection, data= [ 0x02 ] 2
                            Logical
            Item(Local ): Usage Minimum, data= [ 0x01 ] 1
                            Alphanumeric Display
            Item(Local ): Usage Maximum, data= [ 0x66 ] 102
                            (null)
            Item(Global): Logical Minimum, data= [ 0x00 ] 0
            Item(Global): Logical Maximum, data= [ 0x01 ] 1
            Item(Global): Report ID, data= [ 0x02 ] 2
            Item(Global): Report Size, data= [ 0x08 ] 8
            Item(Global): Report Count, data= [ 0x0b ] 11
            Item(Local ): Usage, data= [ 0x2c ] 44
                            Display Data
            Item(Main  ): Output, data= [ 0x02 ] 2
                            Data Variable Absolute No_Wrap Linear
                            Preferred_State No_Null_Position Non_Volatile Bitfield
            Item(Main  ): End Collection, data=none
            Item(Main  ): End Collection, data=none
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x000a  1x 10 bytes
        bInterval              10
Device Status:     0x0000
  (Bus Powered)
```

**Comment by rsm-gh** [comment]
Hello @KickinWingAnimalDoctor, I've spitted your bug request in two since you're actually [having problems because of Ubuntu](https://github.com/rsm-gh/alienware-kbl/issues/15).

I moved your bug report in to the previous link, and I added the support to the Alienware 13R3 in to the testing code. So here we will be able to follow specific problems about the Alienware 13R3 like inverting zones or finding hex blocks, while in the other bug I'll try to fix `alienware-kbl` for Ubuntu !

**Comment by KickinWingAnimalDoctor** [comment]
Thanks for the quick response rsm-gh! I have setup the test branch but when launching the app I encountered this error. Can alienware-kbl run without the daemon? If not I will hold off on submissions until it is functioning in ubuntu 16.04.

$sudo alienware-kbl
Runing without daemon.

(GUI.py:3286): Gtk-WARNING **: Unknown type < define a new column > specified in treemodel liststore_default_blocks

(GUI.py:3286): Gtk-WARNING **: /build/gtk+3.0-PCLr8A/gtk+3.0-3.18.9/./gtk/gtkliststore.c:516: Invalid type (null)

Exception in thread Thread-2:
Traceback (most recent call last):
  File "/usr/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/share/alienware-kbl/GUI.py", line 720, in ILUMINATE_keyboard
    self.controller.Write_Conf()
  File "/usr/share/alienware-kbl/Engine.py", line 179, in Write_Conf
    self.WaitForOk()
  File "/usr/share/alienware-kbl/Engine.py", line 255, in WaitForOk
    self.Get_State()
  File "/usr/share/alienware-kbl/Engine.py", line 271, in Get_State
    msg = self.driver.ReadDevice(request)
  File "/usr/share/alienware-kbl/Engine.py", line 120, in ReadDevice
    msg = self.dev.ctrl_transfer(self.READ_REQUEST_TYPE, self.READ_REQUEST, self.READ_VALUE, self.READ_INDEX, len(msg[0].packet))
  File "/usr/lib/python3/dist-packages/usb/core.py", line 971, in ctrl_transfer
    self.__get_timeout(timeout))
  File "/usr/lib/python3/dist-packages/usb/backend/libusb1.py", line 819, in ctrl_transfer
    timeout))
  File "/usr/lib/python3/dist-packages/usb/backend/libusb1.py", line 552, in _check
    raise USBError(_strerror(ret), ret, _libusb_errno[ret])
  File "/usr/lib/python3/dist-packages/usb/backend/libusb1.py", line 541, in _strerror
    return _lib.libusb_strerror(errcode).decode('utf8')
  File "/usr/lib/python3.5/ctypes/__init__.py", line 360, in __getattr__
    func = self.__getitem__(name)
  File "/usr/lib/python3.5/ctypes/__init__.py", line 365, in __getitem__
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: /usr/share/alienware-kbl/libusb-1.0.so.0: undefined symbol: libusb_strerror

**Comment by rsm-gh** [comment]
A feature of the program is that the GUI is able to run as without the daemon if it is started with root privileges. Sadly you did this, and you're encountering an USB error !

I'll try the software under Ubuntu and I hope I find some solution for this.

**Comment by KickinWingAnimalDoctor** [comment]
Hi rsm-gh,

Good news! I have gotten everything to work. I had to force alienware-kbl to use the libusb version obtained from here: https://github.com/libusb/libusb

After this I had to edit the value in Computers.py.  "DATA_LENGTH = 9" was changed to 12. 

I am still testing colors and other things but the leds come on and off on command. Thanks again for all your help!

**Comment by rsm-gh** [comment]
I'm happy that you've found the right libusb version that works for Ubuntu!

Sadly this creates some problems for the software because it means that I need to pack different libusb libraries, detect the distribution, and maintain the files.

It is an old problem that comes back and it would be great to find a solid solution.

Long time ago I tried to use the libusb version that is officially packaged with the distributions. I know that the library is located at `/usr/lib/libusb-1.0.so.0` after installing libusb, but I haven been able to make it work.

And concerning the fact that you changed the `DATA_LENGTH` variable is pretty interesting because it would be the first time that it needs to be modified. 

Does all the lights work okay? (I mean, beside having inverted zones or stuff like that). I think that modifying the value of this variable may affect other functionalities like the morph, fixed, or blink modes.

If you have any other update don't hesitate to post it because I really appreciate what you've found!

**Comment by KickinWingAnimalDoctor** [comment]
Right now all of the lights work besides the track pad. I also appear to only have two colors (red and green). I will investigate more into this when I have time. As for the DATA_LENGTH value, it was a pure guess. When running the program with it set to 9, libusb would throw a buffer overflow error. I want to decrease this value and see what the minimum is needed to get the program to function. I will post my findings here. Thanks again!

**Comment by Zephyr-D** [comment]
Hi guys,
I've read all above and find it very interesting and helpful.
Will it also be the solution for 15R3? I think they shall be similar things. I'm also running Ubuntu (16.04).
Thank you!

**Comment by rsm-gh** [comment]
Hi @Zephyr-D, 

The problem about using Ubuntu is actually reported [here](https://github.com/rsm-gh/alienware-kbl/issues/15). I'll try to submit something soon !

**Comment by MrMEEE** [comment]
I also have the Alienware 13 R3 and am willing to help out..

**Comment by rsm-gh** [comment]
Hi @MrMEEE , you can then refer to the [F.A.Q.](https://github.com/rsm-gh/alienware-kbl/wiki/F.A.Q.) question [My computer is not supported! What can I do?](https://github.com/rsm-gh/alienware-kbl/wiki/F.A.Q.#my-computer-is-not-supported-what-can-i-do) and start by installing the [testing branch](https://github.com/rsm-gh/alienware-kbl/tree/testing) since I've already added the configuration thanks to @KickinWingAnimalDoctor .

You can then continue with the instructions of the F.A.Q. and report any problem you have! In case you have the same problems than the ones mentioned at the top, you could try the software under other distro with some live USB. I'd recommend Debian and even better ArchLinux since its the distro I use.


Keep me updated and thanks :)

**Comment by sev7enITA** [comment]
Dears, thanks as first and me also I am with the Alienware 13 R3 OLED.

**Comment by MrMEEE** [comment]
@rsm-gh I have tried the testing branch...

The service fails to start:

mj@falcon:~/Download/alienware-kbl-testing$ maj 15 10:29:13 falcon sudo[10889]:       mj : TTY=pts/1 ; PWD=/home/mj/Download/alienware-kbl-testing ; USER=root ; COMMAND=/bin/systemctl restart alienware-kbl
maj 15 10:29:13 falcon sudo[10889]: pam_unix(sudo:session): session opened for user root by (uid=0)
maj 15 10:29:13 falcon systemd[1]: Started alienware-kbl daemon.
maj 15 10:29:13 falcon sudo[10889]: pam_unix(sudo:session): session closed for user root
maj 15 10:29:13 falcon bash[10892]: Traceback (most recent call last):
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 113, in _iluminate_keyboard
maj 15 10:29:13 falcon bash[10892]:     self._indicator_pyro.load_profiles(
maj 15 10:29:13 falcon bash[10892]: AttributeError: 'bool' object has no attribute 'load_profiles'
maj 15 10:29:13 falcon bash[10892]: Traceback (most recent call last):
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 366, in <module>
maj 15 10:29:13 falcon bash[10892]:     ConnectDaemon()
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 41, in __init__
maj 15 10:29:13 falcon bash[10892]:     uri = self.daemon.register(Daemon(self))
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 93, in __init__
maj 15 10:29:13 falcon bash[10892]:     self.set_lights(self._user, self._ccp.get_bool_defval('boot', True))
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 250, in set_lights
maj 15 10:29:13 falcon bash[10892]:     self._iluminate_keyboard()
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Daemon.py", line 132, in _iluminate_keyboard
maj 15 10:29:13 falcon bash[10892]:     self._controller.Write_Conf()
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Engine.py", line 195, in Write_Conf
maj 15 10:29:13 falcon bash[10892]:     self.WaitForOk()
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Engine.py", line 282, in WaitForOk
maj 15 10:29:13 falcon bash[10892]:     self.Get_State()
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Engine.py", line 298, in Get_State
maj 15 10:29:13 falcon bash[10892]:     msg = self.driver.ReadDevice(request)
maj 15 10:29:13 falcon bash[10892]:   File "/usr/share/alienware-kbl/Engine.py", line 135, in ReadDevice
maj 15 10:29:13 falcon bash[10892]:     msg[0].packet))
maj 15 10:29:13 falcon bash[10892]:   File "/usr/lib/python3/dist-packages/usb/core.py", line 1043, in ctrl_transfer
maj 15 10:29:13 falcon bash[10892]:     self.__get_timeout(timeout))
maj 15 10:29:13 falcon bash[10892]:   File "/usr/lib/python3/dist-packages/usb/backend/libusb1.py", line 883, in ctrl_transfer
maj 15 10:29:13 falcon bash[10892]:     timeout))
maj 15 10:29:13 falcon bash[10892]:   File "/usr/lib/python3/dist-packages/usb/backend/libusb1.py", line 595, in _check
maj 15 10:29:13 falcon bash[10892]:     raise USBError(_strerror(ret), ret, _libusb_errno[ret])
maj 15 10:29:13 falcon bash[10892]: usb.core.USBError: [Errno 75] Overflow
maj 15 10:29:13 falcon systemd[1]: alienware-kbl.service: Main process exited, code=exited, status=1/FAILURE
maj 15 10:29:13 falcon systemd[1]: alienware-kbl.service: Unit entered failed state.
maj 15 10:29:13 falcon systemd[1]: alienware-kbl.service: Failed with result 'exit-code'.

**Comment by arcangeli** [comment]
Hi, I've installed Arch on my new 13r3 OLED this morning.
alienware-kbl open the GUI but error appear when trying to apply (after save):
``ERROR from `/usr/share/AlienwareKBL/Bindings.py` on method `_command`:
Command=set_lights, arguments=[root,True]
Traceback (most recent call last):
  File "/usr/share/AlienwareKBL/Bindings.py", line 54, in _command
    response = getattr(self._pyro, command)(*args)
  File "/usr/lib/python3.6/site-packages/Pyro4/core.py", line 187, in __call__
    return self.__send(self.__name, args, kwargs)
  File "/usr/lib/python3.6/site-packages/Pyro4/core.py", line 463, in _pyroInvoke
    data = serializer.deserializeData(msg.data, compressed=msg.flags & message.FLAGS_COMPRESSED)
  File "/usr/lib/python3.6/site-packages/Pyro4/util.py", line 170, in deserializeData
    return self.loads(data)
  File "/usr/lib/python3.6/site-packages/Pyro4/util.py", line 598, in loads
    return self.recreate_classes(serpent.loads(data))
  File "/usr/lib/python3.6/site-packages/Pyro4/util.py", line 415, in recreate_classes
    return self.dict_to_class(literal)
  File "/usr/lib/python3.6/site-packages/Pyro4/util.py", line 617, in dict_to_class
    return super(SerpentSerializer, cls).dict_to_class(data)
  File "/usr/lib/python3.6/site-packages/Pyro4/util.py", line 394, in dict_to_class
    raise errors.SerializeError("unsupported serialized class: " + classname)
Pyro4.errors.SerializeError: unsupported serialized class: usb.core.USBError
``
I've got some results with --block-testing.
[alienware-kbl_a13r3.txt](https://github.com/rsm-gh/alienware-kbl/files/1768125/alienware-kbl_a13r3.txt)

Can't light up the touchpad.

**Comment by rsm-gh** [comment]
@arcangeli thanks for testing the software with arch !

From what I see I can tell you that the bug:

```
Pyro4.errors.SerializeError: unsupported serialized class: usb.core.USBError
``` 
seems to be kinda common and I haven't been able to reproduce it :/. Its a real shame because it complicates me so much to find a fix. Anyways your test tells me that is not because of the Pyro version because since you're in arch you must be using the same than me. I'll try to make more research.

Concerning the colors/settings you joined:
```
[Device found]: Vendor ID: 6268	 Product ID: 1321

DATA_LENGTH = 12

# This work, all lights goes off.
[Command]: Lights off

# Colors are always messed up. Full red (for the two colors) give something as a pinkish orange. But colors are coherent on all parts.
# At first, i was thinking that there are some inverted color (like GRB in place of RGB) Will make more test at night soon.

# 4 parts for the keyboard.
# This is the full-right part of the keyboard
[TEST]: block: 1	 hex: 0x1	 mode:blink	 speed:1	 color1:#204a87	 color2: #204a87

# This is the center-right part of the keyboard
TEST]: block: 2	 hex: 0x2	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# This is the center-left part of the keyboard
[TEST]: block: 4	 hex: 0x4	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# This is the full-left part of the keyboard
[TEST]: block: 8	 hex: 0x8	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# End Keyboard

# This is the logo behind the screen.
TEST]: block: 32	 hex: 0x20	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# This is the text logo under the screen
[TEST]: block: 64	 hex: 0x40	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# This is the power-alien-button
[TEST]: block: 256	 hex: 0x100	 mode:fixed	 speed:1	 color1:#ef2929	 color2: #ef2929

# That's all :)
```

I wonder if the colors are messed up because of the `DATA_LENGTH = 12` variable. I think that it would worth to try the block testing window without modifying that variable and check if there's an USB.core.error.

Please keep me updated ! 

Regards,
rsm

**Comment by arcangeli** [comment]
Some color test.
[alienware-kbl_a13r3_2.txt](https://github.com/rsm-gh/alienware-kbl/files/1770311/alienware-kbl_a13r3_2.txt)
Only work with DATA_LENGTH=12.
Thanks for your help.

**Comment by MrMEEE** [comment]
@arcangeli Did you get the touchpad to light up???

**Comment by arcangeli** [comment]
No ;(

**Comment by Soulsuke** [comment]
About the touchpad, there is an option within the BIOS (forgot its name, should be in the advanced tab) which can be set to either keep the touchpad light always on, or to make it turn off automatically after 10 seconds of no use... Sadly, this last setting makes it not to light up at all under linux (so far)... Could this be the issue??

**Comment by gggauravgandhi** [comment]
I am not able to make it work for Alienware R313 on Ubuntu 18.04

`Pyro4.errors.SerializeError: unsupported serialized class: usb.core.USBError`

**Comment by StephanX** [comment]
Hi there!  So, I have a 13R3, `Device: ID 187c:0529 Alienware Corporation` on arch-linux.  With `libusb 1.0.22-1`, block-testing is able to toggle lights on and off, so long as DATA_LENGTH is flipped to 12.  However, it seems like no matter what I do to the source code (whether I download and recursively sed, or recursively sed/grep for DATA_LENGTH in /usr/lib/python*), the DATA_LENGTH seems hard coded to 9 when I try running the akbl client.

Separately, when activating region 128, or 1048476 in block-testing, the touchpad lights up.  I do find  that the specified colors usually don't end up matching the result colors, but I'm really more interested in just having the keyboard backlight on, and not morphing (which, by default, seems to happen no matter what, when I set in windows.)

Am happy to help test or investigate any other changes, but the DATA_LENGTH default really does have me stumped.  Thanks!

**Comment by StephanX** [comment]
Not to detract from this project, but after a few hours of trial and error, I found https://github.com/ssteinberg/alienfx_aw13r3 simply works 'out-of-the-box' for the 13R3.  Maybe there's some secret sauce to take from it?  It's also a python based project.

**Comment by rsm-gh** [comment]
Hello @StephanX, 
The `DATA_LENGTH` variable is not hard coded, [its value can be modified in the configuration files](https://github.com/rsm-gh/akbl/blob/a6e9819e429cd97ad54076a4c715f7ae6af08824/usr/share/AKBL/computers/Alienware13R3.ini#L57). Then, if you modify this variable, the  Data Length is not the one provided by default so it is normal that the colors won't match. 

This problem is currently affecting many computers and I'd be happy to fix it but it's complicated since it doesn't happen in my computer. I'll try take a look to the other project that you sent but without the usb logs I'll probably won' t be able of debugging.

**Comment by StephanX** [comment]
I much appreciate the reply!  That's the config file I modified, but something just didn't stick.  If you're willing to provide me with instructions (either here, or directly), I'd be happy to dump usb logs from my machine.

**Comment by rsm-gh** [comment]
@StephanX  I just committed the [bug fix](https://github.com/rsm-gh/akbl/commit/07a42ee7f071c55b52305a31641043f7193dc19d) that was stopping akbl to update your modifications in the configuration file. It should be okay now.

And concerning the other project, well it may take time since I installed it for doing some testing and debugging and it doesn't work with my M14XR1. I'll keep looking for curiosity but It may not help me :s
